src/services/postgres_rfqt_firm_quote_validator.ts

```diff
@@ -0,0 +1,188 @@
+import { BigNumber, RfqtFirmQuoteValidator, SignedOrder } from '@0x/asset-swapper';
+import { assetDataUtils, ERC20AssetData } from '@0x/order-utils';
+import { In } from 'typeorm';
+import { Repository } from 'typeorm/repository/Repository';
+
+import { ONE_MINUTE_MS, ZERO } from '../constants';
+import { MakerBalanceChainCacheEntity } from '../entities/MakerBalanceChainCacheEntity';
+import { logger } from '../logger';
+
+const THRESHOLD_CACHE_EXPIRED_MS = ONE_MINUTE_MS * 2;
+
+export class PostgresRfqtFirmQuoteValidator implements RfqtFirmQuoteValidator {
+    private readonly _chainCacheRepository: Repository<MakerBalanceChainCacheEntity>;
+    private readonly _cacheExpiryThresholdMs: number;
+
+    private static _calculateTakerFillableAmountsFromQuotes(
+        quotes: SignedOrder[],
+        makerLookup: { [key: string]: BigNumber },
+    ): { makerAddressesToAddToCache: string[]; takerFillableAmounts: BigNumber[] } {
+        const makerAddressesToAddToCacheSet: Set<string> = new Set();
+        const takerFillableAmounts = quotes.map(quote => {
+            const makerTokenBalanceForMaker: BigNumber | undefined = makerLookup[quote.makerAddress];
+
+            // TODO: Add Prometheus hooks
+            if (makerTokenBalanceForMaker === undefined) {
+                makerAddressesToAddToCacheSet.add(quote.makerAddress);
+                return quote.takerAssetAmount;
```

Comment:
is it worth taking the hit and pulling the balance on-chain?

Reply:
discussed, will talk to Blake on Monday